File: 6/6.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-


class Solution(object):
	def convert(self, s, nRows):
		""" 
		: type s:str
		: type nRows:int 
		: rtype: str
		"""
		nLen = len(s)
		if nLen == 0:
		  print 'string is empty'
		  return ''
		if nRows == 1 or nRows >= nLen:
		  return s

		nStepSum = (nRows - 1) * 2
		nStep1 = nStep2 = 0

		result = ''
		for i in range(0, nRows):
			if i == 0 or i == (nRows - 1):
				nStep1 = nStep2 = nStepSum
			else:
				nStep2 = i * 2
				nStep1 = nStepSum - nStep2
				
			j = i	
			while j < nLen:
				result += s[j]
				j += nStep1
				if j < nLen:
					result += s[j]
					j += nStep2
				else:
					break;
					
		return result	


def main():
	slt=Solution()
	print slt.convert('PAYPALISHIRING', 1)
	print slt.convert('PAYPALISHIRING', 2)
	print slt.convert('PAYPALISHIRING', 3)
	print slt.convert('PAYPALISHIRING', 4)
	print slt.convert('PAYPALISHIRING', 5)
	print slt.convert('PAYPALISHIRING', 15)


if __name__ == '__main__':
	main()

File: 6/6.c

/*

   The string "PAYPALISHIRING" is written in a zigzag pattern on a given number of rows like this: (you may want to display this pattern in a fixed font for better legibility)

   P   A   H   N
   A P L S I I G
   Y   I   R

   And then read line by line: "PAHNAPLSIIGYIR"

   Write the code that will take a string and make this conversion given a number of rows:

   string convert(string text, int nRows);

   convert("PAYPALISHIRING", 3) should return "PAHNAPLSIIGYIR". 


 */


#include <stdio.h>
#include <string.h>
#include <malloc.h>

char * convert( char * s, int nRows)
{
	if(!s) return NULL;

	int nLen = strlen(s);
	if (nRows == 1 || nRows >= nLen) return s;

	char * buff = (char *)malloc(nLen + 1);
	if(!buff) 
	{
		return NULL;	
	}
	else
	{
		buff[nLen] = '\0';
	}

	int nStepSum = (nRows - 1) * 2;
	const char * p = NULL;
	const char * pStrEnd = s + nLen;
	char * q = buff;
	int nStep1, nStep2;
	nStep1 = nStep2 = 0;
	for (int i = 0; i < nRows ; i++) 
	{
		if (i == 0 || i == (nRows - 1))
		{
			nStep1 = nStep2 = nStepSum;
		}
		else
		{
			nStep2 = i * 2;
			nStep1 = nStepSum - nStep2;
		}

		p = s + i;
		while(p < pStrEnd) 
		{
			*q++ = *p;
			p += nStep1;
			if (p < pStrEnd) 
			{
				*q++ = *p;
				p += nStep2;
			}
			else
			{
				break;
			}
		}

	}

	strcpy(s, buff);
	free(buff);
	return s;
}

int main()
{
	char s3[] = "PAYPALISHIRING";
	printf("%s\n", convert(s3, 1));
	char s2[] = "PAYPALISHIRING";
	printf("%s\n", convert(s2, 2));
	char s1[] = "PAYPALISHIRING";
	printf("%s\n", convert(s1, 3));
	char s4[] = "PAYPALISHIRING";
	printf("%s\n", convert(s4, 4));
	char s5[] = "PAYPALISHIRING";
	printf("%s\n", convert(s5, 5));
	char s6[] = "PAYPALISHIRING";
	printf("%s\n", convert(s6, 15));
	return 0;
}
